daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala

```diff
@@ -27,216 +27,434 @@ import org.apache.daffodil.schema.annotation.props.gen.OccursCountKind
+          ssp match {
+            // case AnyEmpty => NonPositional
+            case _ => Positional
+          }
+        }
+      case eb: ElementBase if (eb.isArray || eb.isOptional) => {
+        import OccursCountKind._
+        eb.occursCountKind match {
+          case Fixed => Positional
+          case Expression => Positional
+          case Parsed => NonPositional
+          case StopValue => NonPositional
+          case Implicit => {
+            val UNB = -1
+            (eb.isPotentiallyTrailing,
+              this.isDeclaredLast,
+              sq.separatorSuppressionPolicy,
+              eb.maxOccurs) match {
+                case (false, _, AnyEmpty, _) => NonPositional
+                case (false, _, _, UNB) if !eb.isLastDeclaredRepresentedInSequence =>
+                  Assert.invariantFailed("Should be SDE found elsewhere.")
+                case (false, _, _, _) => Positional
+                case (true, _, Never, UNB) => Assert.invariantFailed("Should be SDE found elsewhere.")
+                case (true, false, AnyEmpty, UNB) => NonPositional
+                case (true, false, TrailingEmpty, UNB) if !eb.isLastDeclaredRepresentedInSequence =>
+                  Assert.invariantFailed("Should be SDE found elsewhere.")
+                case (true, false, TrailingEmptyStrict, UNB) if !eb.isLastDeclaredRepresentedInSequence =>
+                  Assert.invariantFailed("Should be SDE found elsewhere.")
+                case (true, _, Never, _) => Positional
+                case (true, _, AnyEmpty, _) => NonPositional
+                case (true, _, TrailingEmpty, _) => PositionalTrailingLax
+                case (true, _, TrailingEmptyStrict, _) => PositionalTrailingStrict
+              }
+          }
         }
       }
-      case m: ModelGroup => {
-        new NeverZeroLengthDetector
+      case eb: ElementBase if eb.isScalar => {
+        if (child.isPotentiallyTrailing) {
+          ssp match {
+            case AnyEmpty => NonPositional
+            case TrailingEmpty => PositionalTrailingLax
+            case TrailingEmptyStrict => PositionalTrailingStrict
+            case Never => Positional
+          }
+        } else {
+          ssp match {
+            // case AnyEmpty => NonPositional
+            case _ => Positional
+          }
+        }
       }
+      case _ => Positional // obscure cases like maxOccurs 0 with lengthKind 'implicit'
+    }
+  }
+  protected final def sscb = separatedSequenceChildBehavior
+
+  final protected def isDeclaredLast: Boolean = {
+    child.isLastDeclaredRepresentedInSequence
+  }
+
+  final protected def isPositional: Boolean = {
+    separatedSequenceChildBehavior match {
+      case _: SeparatedSequenceChildBehavior.PositionalLike => true
+      case _ => false
     }
-    result
   }
 
   /**
-   * Used when unparsing to determine which of several separator suppression
-   * algorithms is required.
+   * Combines static knowledge of whether the term can
```

Comment:
This block comment no longer belongs here if the zeroLengthDetector has been moved from here.